File: CSC310-Cpp/Module11-Graph-Algorithms/debug02/topological_sort.cpp
// topological_sort.cpp
// Topological sort of a DAG by depth-first search (CLRS 20.4). The example
// is the getting-dressed graph: an edge u -> v means u must be put on before
// v. After producing an order, the driver verifies that every edge points
// forward in it.
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

const std::vector<std::string> names = {"undershorts", "pants", "belt", "shirt", "tie",
                                        "jacket", "socks", "shoes", "watch"};

void dfsVisit(const std::vector<std::vector<int>>& adj, int u, std::vector<bool>& visited,
              std::vector<int>& order) {
    visited[u] = true;
    order.push_back(u);
    for (int v : adj[u]) {
        if (!visited[v]) {
            dfsVisit(adj, v, visited, order);
        }
    }
}

std::vector<int> topologicalSort(const std::vector<std::vector<int>>& adj) {
    const int n = static_cast<int>(adj.size());
    std::vector<bool> visited(n, false);
    std::vector<int> order;
    for (int u = 0; u < n; ++u) {
        if (!visited[u]) {
            dfsVisit(adj, u, visited, order);
        }
    }
    std::reverse(order.begin(), order.end());
    return order;
}

int main() {
    const int n = static_cast<int>(names.size());
    std::vector<std::vector<int>> adj(n);
    std::vector<std::pair<int, int>> edges = {
        {0, 1}, {0, 7}, {1, 2}, {1, 7}, {2, 5}, {3, 2}, {3, 4}, {4, 5}, {6, 7}};
    for (auto [u, v] : edges) {
        adj[u].push_back(v);
    }

    std::vector<int> order = topologicalSort(adj);
    std::vector<int> position(n);
    std::cout << "order:";
    for (int i = 0; i < n; ++i) {
        position[order[i]] = i;
        std::cout << ' ' << names[order[i]];
    }
    std::cout << '\n';

    int violations = 0;
    for (auto [u, v] : edges) {
        if (position[u] > position[v]) {
            std::cout << "  violated: " << names[u] << " -> " << names[v] << '\n';
            ++violations;
        }
    }
    std::cout << (violations == 0 ? "valid topological order" : "NOT a valid topological order")
              << '\n';
    return 0;
}

File: CSC310-Python/Module11-Graph-Algorithms/debug02/topological_sort.py
"""Topological sort by depth-first search (CLRS Section 20.4).

topological_sort(adj) returns a linear ordering of the vertices of a DAG
such that every edge (u, v) has u before v.  The main block orders the
getting-dressed graph of CLRS Section 20.4 and verifies the ordering
against every edge.
"""


def topological_sort(adj):
    WHITE, GRAY, BLACK = 0, 1, 2
    color = {v: WHITE for v in adj}
    finished = []

    def visit(u):
        color[u] = GRAY
        finished.append(u)
        for v in adj[u]:
            if color[v] == WHITE:
                visit(v)
        color[u] = BLACK

    for u in adj:
        if color[u] == WHITE:
            visit(u)
    return finished


def violations(adj, order):
    position = {v: i for i, v in enumerate(order)}
    return [(u, v) for u in adj for v in adj[u] if position[u] > position[v]]


def main():
    adj = {
        "undershorts": ["pants", "shoes"],
        "pants": ["belt", "shoes"],
        "belt": ["jacket"],
        "shirt": ["belt", "tie"],
        "tie": ["jacket"],
        "socks": ["shoes"],
        "watch": [],
        "shoes": [],
        "jacket": [],
    }
    order = topological_sort(adj)
    print("order:")
    for i, v in enumerate(order, 1):
        print(f"  {i}. {v}")
    bad = violations(adj, order)
    print("edges out of order:", bad if bad else "none")
    print("valid topological order:", not bad)


if __name__ == "__main__":
    main()
